bakery/src/scripts/copy-resources-s3.py

```diff
@@ -0,0 +1,269 @@
+                    aws_secret=aws_secret,
+                    bucket=bucket,
+                    resource=resource,
+                    disable_check=disable_deep_folder_check)
+            )
+        for future in concurrent.futures.as_completed(check_futures):
+            try:
+                resource = future.result()
+            except Exception as e:
+                print(e)    # print error from ThreadPoolExecutorStackTraced
+                halt_all_threads(executor)
+                sys.exit(1)
+            finally:
+                check_futures.remove(future)
+                del future
+            try:
+                if resource is not None:
+                    upload_resources.append(resource)
+                    if not disable_deep_folder_check:
+                        print('.', end='', flush=True)
+                else:
+                    if not disable_deep_folder_check:
+                        print('x', end='', flush=True)
+            except:
+                halt_all_threads(executor)
+                raise
+    if disable_deep_folder_check:
+        print('- quick checked (destination S3 folder empty or non existing)',
+              end='', flush=True)
+    print()
+    print('{} resources need uploading.'.format(len(upload_resources)))
+    elapsed = (timer() - start)
+    print('Time it took to check: {}s'.format(elapsed))
+
+    # upload to s3 (with ThreadPoolExecutor)
+    start = timer()
+    upload_count = 0
+    upload_futures = []
+    with ThreadPoolExecutorStackTraced(max_workers=MAX_THREAD_UPLOAD_S3) as executor:
+        print('Uploading to S3 ', end='', flush=True)
+        for resource in upload_resources:
+            # upload metadata first (because this is not checked)
+            upload_futures.append(
+                executor.submit(
+                    upload_s3,
+                    aws_key=aws_key,
+                    aws_secret=aws_secret,
+                    filename=resource['input_metadata_file'],
+                    bucket=bucket,
+                    key=resource['output_s3_metadata'],
+                    content_type='application/json')
+            )
+            # upload resource file last (existence/md5 is checked)
+            upload_futures.append(
+                executor.submit(
+                    upload_s3,
+                    aws_key=aws_key,
+                    aws_secret=aws_secret,
+                    filename=resource['input_file'],
+                    bucket=bucket,
+                    key=resource['output_s3'],
+                    content_type=resource['mime_type'])
+            )
+        for future in concurrent.futures.as_completed(upload_futures):
+            try:
+                result = future.result()
+            except Exception as e:
+                print(e)    # print error from ThreadPoolExecutorStackTraced
+                halt_all_threads(executor)
+                sys.exit(1)
+            finally:
```

Comment:
Why is this finally block necessary?

Reply:
You are right, thx. I think finally is not good. I will better move the lines of the finally block into the second try block (which is also better for thread stopping... I'm pretty sure now that  threads are not stopped on errors in the final block)